store/tikv/region_cache.go

```diff
@@ -462,119 +574,124 @@ func (c *RegionCache) loadRegionByID(bo *Backoffer, regionID uint64) (*Region, e
+		workPeer = cachedPeer
+		return
 	}
-	return addr, nil
-}
-
-// ClearStoreByID clears store from cache with storeID.
-func (c *RegionCache) ClearStoreByID(id uint64) {
-	c.storeMu.Lock()
-	defer c.storeMu.Unlock()
-	delete(c.storeMu.stores, id)
-}
 
-func (c *RegionCache) loadStoreAddr(bo *Backoffer, id uint64) (string, error) {
+	// try round-robin find & switch to other peers when old leader meet error.
+	newIdx := -1
+	storeNum := len(region.stores)
+	i := (cachedIdx + 1) % storeNum
+	start := i
 	for {
-		store, err := c.pdClient.GetStore(bo.ctx, id)
-		if err != nil {
-			tikvRegionCacheCounterWithGetStoreError.Inc()
-		} else {
-			tikvRegionCacheCounterWithGetStoreOK.Inc()
-		}
-		if err != nil {
-			if errors.Cause(err) == context.Canceled {
-				return "", errors.Trace(err)
-			}
-			err = errors.Errorf("loadStore from PD failed, id: %d, err: %v", id, err)
-			if err = bo.Backoff(BoPDRPC, err); err != nil {
-				return "", errors.Trace(err)
-			}
-			continue
+		store := region.stores[i]
+		if store.Available(ts) {
+			newIdx = i
+			break
 		}
-		if store == nil {
-			return "", nil
+		i = (i + 1) % storeNum
+		if i == start {
+			break
 		}
-		return store.GetAddress(), nil
 	}
+	if newIdx < 0 {
+		return
+	}
+	if !atomic.CompareAndSwapInt32(&region.workStoreIdx, int32(cachedIdx), int32(newIdx)) {
+		goto retry
+	}
+	if int32(newIdx) == atomic.LoadInt32(&region.syncStoreIdx) {
+		region.scheduleReload()
+	}
+	workStore = region.stores[newIdx]
+	workPeer = region.meta.Peers[newIdx]
+	return
 }
 
-// DropStoreOnSendRequestFail is used for clearing cache when a tikv server does not respond.
-func (c *RegionCache) DropStoreOnSendRequestFail(ctx *RPCContext, err error) {
-	// We need to drop the store only when the request is the first one failed on this store.
-	// Because too many concurrently requests trying to drop the store will be blocked on the lock.
-	failedRegionID := ctx.Region
-	failedStoreID := ctx.Peer.StoreId
-	c.mu.Lock()
-	_, ok := c.mu.regions[failedRegionID]
-	if !ok {
-		// The failed region is dropped already by another request, we don't need to iterate the regions
-		// and find regions on the failed store to drop.
-		c.mu.Unlock()
-		return
+// hasAvailableStore checks whether region has available store.
+// different to `routeStoreInRegion` just check and never change work store or peer.
+func (c *RegionCache) hasAvailableStore(region *Region, ts int64) bool {
+	if len(region.stores) == 0 {
+		return false
 	}
-	for id, r := range c.mu.regions {
-		if r.region.peer.GetStoreId() == failedStoreID {
-			c.dropRegionFromCache(id)
+	for _, store := range region.stores {
+		if store.Available(ts) {
+			return true
 		}
 	}
-	c.mu.Unlock()
+	return false
+}
 
-	// Store's meta may be out of date.
-	var failedStoreAddr string
+func (c *RegionCache) getStoreByStoreID(storeID uint64) (store *Store) {
+	var ok bool
 	c.storeMu.Lock()
-	store, ok := c.storeMu.stores[failedStoreID]
+	store, ok = c.storeMu.stores[storeID]
```

Comment:
`defer c.storeMu.Unlock()`?

Reply:
this logic should not trigger panic, so here just want to avoid defer cost conveniently :D